File: 2. Structural Design Patterns/Bridge/c++/BridgeTemplate.cpp
#include <iostream>
using namespace std;

// Implementation Interface
class Implementation {
public:
    virtual void operationImpl() = 0;
    virtual ~Implementation() {}
};

// Concrete Implementations
class ConcreteImplA : public Implementation {
public:
    void operationImpl() override {
        cout << "ConcreteImplA operation\n";
    }
};

class ConcreteImplB : public Implementation {
public:
    void operationImpl() override {
        cout << "ConcreteImplB operation\n";
    }
};

// Abstraction
class Abstraction {
protected:
    Implementation* impl;
public:
    Abstraction(Implementation* impl) : impl(impl) {}
    virtual void operation() {
        impl->operationImpl();
    }
    virtual ~Abstraction() {}
};

// Refined Abstraction
class RefinedAbstraction : public Abstraction {
public:
    RefinedAbstraction(Implementation* impl) : Abstraction(impl) {}
    void operation() override {
        cout << "RefinedAbstraction delegates to ";
        impl->operationImpl();
    }
};

File: 2. Structural Design Patterns/Bridge/Python/BridgeTemplate.py
# Bridge Pattern – Python Template

from abc import ABC, abstractmethod

class Implementation(ABC):
    @abstractmethod
    def operation_impl(self):
        pass

class ConcreteImplementationA(Implementation):
    def operation_impl(self):
        return "ConcreteImplementationA operation"

class ConcreteImplementationB(Implementation):
    def operation_impl(self):
        return "ConcreteImplementationB operation"

class Abstraction:
    def __init__(self, implementation):
        self.implementation = implementation
    
    def operation(self):
        return f"Abstraction: {self.implementation.operation_impl()}"

class ExtendedAbstraction(Abstraction):
    def extended_operation(self):
        return f"ExtendedAbstraction: {self.implementation.operation_impl()}"

def main():
    impl_a = ConcreteImplementationA()
    abstraction = Abstraction(impl_a)
    print(abstraction.operation())
    
    impl_b = ConcreteImplementationB()
    extended = ExtendedAbstraction(impl_b)
    print(extended.extended_operation())

if __name__ == "__main__":
    main()
